assets/test/e2e/test.js

```diff
@@ -33,9 +33,100 @@ describe('', function() {
+    browser.wait(protractor.ExpectedConditions.elementToBeClickable(nextButton), 2000);
+    return nextButton.click();
+  }
+
+  var requestCreateFromTemplate = function(projectName, templateName) {
+    var uri = '/project/' + projectName + '/create';
+    expect(browser.getCurrentUrl()).toContain(uri);
+    var template = element(by.cssContainingText('.catalog h3 > a', templateName));
+    expect(template.isPresent()).toBe(true);
+    return template.click();
+  }
+
+  var attachBuilderImageToSource = function(projectName, builderImageName) {
+    var uri = '/project/' + projectName + '/catalog/images';
+    expect(browser.getCurrentUrl()).toContain(uri);
+    expect(element(by.cssContainingText('h1', "Select a builder image")).isPresent()).toBe(true);
+    var builderImageLink = element(by.cssContainingText('div > h3 > a', builderImageName));
+    expect(builderImageLink.isPresent()).toBe(true);
+    return builderImageLink.click();
+  }
+
+  var createFromSourceSummary = function(projectName, builderImageName, appName) {
+    var uri = '/project/' + projectName + '/create/fromimage';
+    expect(browser.getCurrentUrl()).toContain(uri);
+    expect(element(by.css('.create-from-image h1')).getText()).toEqual(builderImageName);
+    expect(element(by.cssContainingText('div > h2', "Name")).isPresent()).toBe(true);
+    expect(element(by.cssContainingText('div > h2', "Routing")).isPresent()).toBe(true);
+    expect(element(by.cssContainingText('div > h2', "Deployment Configuration")).isPresent()).toBe(true);
+    expect(element(by.cssContainingText('div > h2', "Build Configuration")).isPresent()).toBe(true);
+    expect(element(by.cssContainingText('div > h2', "Scaling")).isPresent()).toBe(true);
+    expect(element(by.cssContainingText('div > h2', "Labels")).isPresent()).toBe(true);
+    element(by.name('appname')).clear().sendKeys(appName);
+    return element(by.buttonText("Create")).click();
+  }
+
+  var createFromTemplateSummary = function(projectName, templateName, parameterNames, labelNames) {
+    var uri = '/project/' + projectName + '/create/fromtemplate';
+    expect(browser.getCurrentUrl()).toContain(uri);
+    expect(element(by.css('.create-from-template h1')).getText()).toEqual(templateName);
+    expect(element(by.cssContainingText('div > h2', "Images")).isPresent()).toBe(true);
+    expect(element(by.cssContainingText('div > h2', "Parameters")).isPresent()).toBe(true);
+    expect(element(by.cssContainingText('div > h2', "Labels")).isPresent()).toBe(true);
+    if (parameterNames) {
+      for (i = 0; i < parameterNames.length; i++) {
+        expect(element(by.cssContainingText('.env-variable-list label.key', parameterNames[i])).isPresent()).toBe(true);
+      }
+    }
+    if (labelNames) {
+      for (i = 0; i < labelNames.length; i++) {
+        expect(element(by.cssContainingText('.label-list span.key', labelNames[i])).isPresent()).toBe(true);
+      }
+    }
+    return element(by.buttonText("Create")).click();
```

Comment:
Yes, it's a promise in the return. My main idea around this design is to allow value propagation and chaining as suggested in WebDriver promises[1], by the use of `then()` which is an observer invoked after the promise is resolved. This allow these functions to be reused by several tests in a readable manner, e.g.:

```
goToIndex().then(login).then(submitStuff).then(function() { loadThisParm(parm); }).then(somethingElse);
```
That makes the promise chain to execute in order. Errors are exceptions that get propagated through the chain, so you can catch them in the end and handle as desired. In the example above, you could either have a try/catch around the entire chain, of a final observer that handles the error:

```
(...).then(null, function(err) { });
```

But of course it's just an idea, I'm open to change this. Do you think returning success/failure would be a better design?

[1] https://github.com/SeleniumHQ/selenium/wiki/WebDriverJs#promises